light/detector.go

```diff
@@ -0,0 +1,223 @@
+package light
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/tendermint/tendermint/light/provider"
+	"github.com/tendermint/tendermint/types"
+)
+
+// The detector component of the light client detect and handles attacks on the light client.
+// More info here:
+// tendermint/docs/architecture/adr-047-handling-evidence-from-light-client.md
+
+// detectDivergence is a second wall of defense for the light client and is used
+// only in the case of skipping verification which employs the trust level mechanism.
+//
+// It takes the target verified header and compares it with the headers of a set of
+// witness providers that the light client is connected to. If a conflicting header
+// is returned it verifies and examines the conflicting header against the verified
+// trace that was produced from the primary. If successful it produces two sets of evidence
+// and sends them to the opposite provider before halting.
+//
+// If there are no conflictinge headers, the light client deems the verified target header
+// trusted and saves it to the trusted store.
+func (c *Client) detectDivergence(primaryTrace []*types.LightBlock, now time.Time) error {
+	var (
+		headerMatched      bool
+		lastVerifiedHeader = primaryTrace[len(primaryTrace)-1].SignedHeader
```

Comment:
if primaryTrace is empty, this will panic

Reply:
but primary trace should never be empty. We initialize the trace like this:
```
trace         = []*types.LightBlock{trustedBlock}
```
in `verifySkipping` before we append to it and finally pass it on to `detectDivergence`